datafusion_ray/core.py

```diff
@@ -0,0 +1,327 @@
+        debug("done stage all done")
+
+    async def new_stage(
+        self,
+        stage_id: int,
+        plan_bytes: bytes,
+        shadow_partition=None,
+    ):
+        stage_key = (stage_id, shadow_partition)
+        try:
+
+            debug(f"creating new stage {stage_key} from bytes {len(plan_bytes)}")
+            stage = RayStage.options(
+                name=f"Stage: {stage_key}, query_id:{self.query_id}",
+            ).remote(
+                stage_id,
+                plan_bytes,
+                shadow_partition,
+            )
+            self.stages[stage_key] = stage
+            self.stages_started.append(stage.start_up.remote())
+
+        except Exception as e:
+            error(
+                f"RayQueryCoordinator[{self.query_id}] Unhandled Exception in new stage! {e}"
+            )
+            raise e
+
+    async def wait_for_stages_ready(self):
+        # TODO: signal our doneness instead of loop
+        while not self.stages_ready:
+            await asyncio.sleep(0.1)
+            debug("waiting for stages to be ready")
+
+    async def ensure_stages_ready(self):
+        if not self.stages_ready:
+            ray.wait(self.stages_started, num_returns=len(self.stages_started))
+            await self.sort_out_addresses()
+            self.stages_ready = True
+        return self.stages_ready
+
+    async def get_stage_addrs(self) -> dict[int, list[str]]:
+        await self.wait_for_stages_ready()
+        return self.stage_addrs
+
+    async def sort_out_addresses(self):
+        for stage_key, stage in self.stages.items():
+            stage_id, shadow_partition = stage_key
+            debug(f" getting stage addr for {stage_id},{shadow_partition}")
+            self.stage_addrs[stage_id].append(await stage.addr.remote())
+
+        debug(f"stage_addrs: {self.stage_addrs}")
+        # now update all the stages with the addresses of peers such
+        # that they can contact their child stages
+        refs = []
+        for stage_key, stage in self.stages.items():
+            refs.append(stage.set_stage_addrs.remote(self.stage_addrs))
+
+        ray.wait(refs, num_returns=len(refs))
+
+    async def serve(self):
+        await self.ensure_stages_ready()
+        info("running stages")
+        try:
+            for stage_key, stage in self.stages.items():
+                info(f"starting serving of stage {stage_key}")
+                stage.serve.remote()
+
+        except Exception as e:
+            error(
+                f"RayQueryCoordinator[{self.query_id}] Unhandled Exception in run stages! {e}"
+            )
+            raise e
+
+
+@ray.remote(num_cpus=0)
+class RayStage:
+    def __init__(
+        self,
+        stage_id: str,
+        plan_bytes: bytes,
+        shadow_partition=None,
+    ):
+
+        from datafusion_ray._datafusion_ray_internal import StageService
+
+        self.shadow_partition = shadow_partition
+        shadow = (
```

Comment:
We can probably compute this in the exception handler it's only used for error handling?